swarm/netsim/simulation.go

```diff
@@ -0,0 +1,262 @@
+func New(services map[string]ServiceFunc, opts *SimulationOptions) (s *Simulation) {
+	s = &Simulation{
+		buckets: make(map[discover.NodeID]*sync.Map),
+		done:    make(chan struct{}),
+	}
+
+	adapterServices := make(map[string]adapters.ServiceFunc, len(services))
+	for name, serviceFunc := range services {
+		s.serviceNames = append(s.serviceNames, name)
+		adapterServices[name] = func(ctx *adapters.ServiceContext) (node.Service, error) {
+			b := new(sync.Map)
+			service, cleanup, err := serviceFunc(ctx, b)
+			if err != nil {
+				return nil, err
+			}
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			if cleanup != nil {
+				s.cleanupFuncs = append(s.cleanupFuncs, cleanup)
+			}
+			s.buckets[ctx.Config.ID] = b
+			return service, nil
+		}
+	}
+
+	s.Net = simulations.NewNetwork(
+		adapters.NewSimAdapter(adapterServices),
+		&simulations.NetworkConfig{ID: "0"},
+	)
+
+	if opts == nil {
+		opts = new(SimulationOptions)
+	}
+	if opts.WithHTTP {
+		s.initHTTPServer(opts)
+	}
+	return s
+}
+
+func (s *Simulation) initHTTPServer(opts *SimulationOptions) {
+	//assign default port if nothing provided
+	if opts.HTTPSimPort == "" {
+		opts.HTTPSimPort = DefaultHTTPSimPort
+	}
+	log.Info(fmt.Sprintf("Initializing simulation server on 0.0.0.0:%s...", opts.HTTPSimPort))
+	//initialize the HTTP server
+	s.handler = simulations.NewServer(s.Net)
+	s.runC = make(chan struct{})
+	//add swarm specific routes to the HTTP server
+	s.addSimulationRoutes()
+	s.httpSrv = &http.Server{
+		Addr:    fmt.Sprintf(":%s", opts.HTTPSimPort),
+		Handler: s.handler,
+	}
+}
+
+func (s *Simulation) startHTTPServer(ctx context.Context) error {
+	//start the HTTP server
+	if s.httpSrv != nil {
+		go s.httpSrv.ListenAndServe()
+		log.Info("Waiting for frontend to be ready...(send POST /runsim to HTTP server)")
+		//wait for the frontend to connect
+		select {
+		case <-s.runC:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		log.Info("Received signal from frontend - starting simulation run.")
+	}
+	return nil
+}
+
+//register additional HTTP routes
+func (s *Simulation) addSimulationRoutes() {
+	s.handler.POST("/runsim", s.RunSimulation)
+}
+
+// StartNetwork starts all nodes in the network
+func (s *Simulation) RunSimulation(w http.ResponseWriter, req *http.Request) {
+	log.Debug("RunSimulation endpoint running")
+	s.runC <- struct{}{}
+	w.WriteHeader(http.StatusOK)
+}
+
+// RunFunc is the function that will be called
+// on Simulation.Run method call.
+type RunFunc func(context.Context, *Simulation) error
+
+// Result is the returned value of Simulation.Run method.
+type Result struct {
+	Duration time.Duration
+	Error    error
+}
+
+// Run calls the RunFunc function while taking care of
+// cancelation provided through the Context.
+func (s *Simulation) Run(ctx context.Context, f RunFunc) (r Result) {
+	//if the option is set to run a HTTP server with the simulation,
+	//init the server and start it
+	start := time.Now()
+	if s.httpSrv != nil {
+		err := s.startHTTPServer(ctx)
```

Comment:
@holisticode we should not start http server in Run method, as we can call Run multiple time and all but first call will not error. Maybe to add WithServer method that will start the servers instead, as @zelig suggested in https://github.com/ethersphere/go-ethereum/pull/769/files#r199888164 comment?